File: 05/sol.py
from collections import deque


def read_file(filename="input.txt"):
    with open(filename, "r") as f:
        return f.read().rstrip()


class Stacks:
    def __init__(self, chunk):
        self.crates = {}
        lines = chunk.split("\n")
        for label in lines[-1].split():
            self.crates[label] = deque()
        for line in reversed(lines[:-1]):
            for i in range(1, len(self.crates)+1):
                j = (i-1)*4+1
                try:
                    char = line[j]
                except:
                    char = " "
                if char.isalpha():
                    self.crates[str(i)].append(char)
        print(self)

    def move(self, fr, to):
        # print("before")
        # print(self)
        # print("move:", fr, to)
        tmp = self.crates[fr].pop()
        self.crates[to].append(tmp)
        # print("after")
        # print(self)

    def moves1(self, count, fr, to):
        for _ in range(count):
            self.move(fr, to)

    def moveall1(self, li):
        for tu in li:
            self.moves1(*tu)

    def moves2(self, count, fr, to):
        tmp = [self.crates[fr].pop() for _ in range(count)]
        self.crates[to].extend(reversed(tmp))

    def moveall2(self, li):
        for tu in li:
            self.moves2(*tu)

    def __str__(self):
        s = ""
        for i in range(1, len(self.crates)+1):
            s += str(i) + ": "
            for c in self.crates[str(i)]:
                s += c
            s += "\n"
        return s.strip()


def parse(input_string):
    chunks = input_string.split("\n\n")
    assert len(chunks) == 2
    stack = Stacks(chunks[0])

    def f(line):
        words = line.split()
        return (int(words[1]), words[3], words[5])

    insts = [f(line) for line in chunks[1].split("\n")]

    return stack, insts


def solve1(stack):
    print(stack)
    return "".join(stack.crates[str(i)][-1] for i in range(1, len(stack.crates)+1))


def solve2(stack):
    print(stack)
    return "".join(stack.crates[str(i)][-1] for i in range(1, len(stack.crates)+1))


if __name__ == "__main__":
    input_string = read_file()
    assert input_string is not None
    stack1, insts1 = parse(input_string)
    stack1.moveall1(insts1)
    res1 = solve1(stack1)
    print(res1)
    stack2, insts2 = parse(input_string)
    stack2.moveall2(insts2)
    res2 = solve2(stack2)
    print(res2)

File: 05/sol.jl
function readstring(filename)
    input = rstrip(read(filename, String))
    @assert !isempty(input) "File Empty"
    return input
end

function parseinput(input)
    crates, moves = split(input, "\n\n")
    crates = reduce(hcat, map(collect, split(crates, "\n")))
    crates = [[c for c in r[begin:end-1] if !isspace(c)] for r in eachrow(crates) if isnumeric(r[end])]
    moves = [split(line)[[2,4,6]] for line in split(moves, "\n")]
    moves = [[parse(Int, i) for i in line] for line in moves]
    moves = [(l[1], (l[2], l[3])) for l in moves]
    return crates, moves
end

function move1(crates::Vector{Vector{Char}}, move::Tuple{Int, Tuple{Int, Int}})
    n, (f, t) = move
    m = crates[f][begin:n]
    crates[f] = crates[f][n+1:end]
    crates[t] = [reverse(m) ; crates[t]]
end

function gettops(crates::Vector{Vector{Char}})::String
    return String([c[begin] for c in crates])
end

function solve1(x)
    crates, moves = x
    for move in moves
        move1(crates, move)
    end
    return gettops(crates)
end

function move2(crates::Vector{Vector{Char}}, move::Tuple{Int, Tuple{Int, Int}})
    n, (f, t) = move
    m = crates[f][begin:n]
    crates[f] = crates[f][n+1:end]
    crates[t] = [m ; crates[t]]
end

function solve2(x)
    crates, moves = x
    for move in moves
        move2(crates, move)
    end
    return gettops(crates)
end

if abspath(PROGRAM_FILE) == @__FILE__
    filename = isempty(ARGS) ? "input.txt" : ARGS[1]
    input = readstring(filename)
    x1 = parseinput(input)
    r1 = solve1(x1)
    println(r1)
    x2 = parseinput(input)
    r2 = solve2(x2)
    println(r2)
end
